[sil] Add flexibility to SILPrinter to allow for full SIL types to printed out in the debugger and programatically without touching the global flag.

gottesmm opened the pull request:
This PR adds flexibility to the SIL printer API such that users can emit full SIL types without touching the global flag. This is implemented by changing SILPrinterContext to take an OptionSet instead of boolean flags and adding a flag to SILPrinterContext that forces the behavior. The default behavior is left alone and is controlled via the global flag.

By doing this, without changing any normal path behavior, we allow for:

1. Compiler developers in the debugger to dump the full types without needing to restart the compilation session to set the global flag.
2. In certain cases when emitting logging specific to an optimization pass, it is useful to emit full types to enable quick triaging. This occurs in RegionAnalysis where we do a quick dump of the entire function before we emit logging to give the user a more global look at the function. It is not intended to be used to get actual output (that is a job for -sil-print-function) but rather just for quick scanning to get a sense of what types/etc are being used. This is especially useful when looking at a large piece of code where types are defined very early in the function (several pages up).

gottesmm committed:
[sil] Change SILPrintContext to take an OptionSet instead of a list of boolean flags.

This will allow for more control of how SIL is printed in logging messages and
also will make it easier to feed flags through SILFunction::dump and friends.

gottesmm committed:
[sil] Add the ability to print out all types in SIL without needing to set the global flag.

This is useful if one wants to dump out state in the debugger or programatically
dump out SIL with all of the types without changing the behavior in the rest of
the compiler.

My intention is to use this to emit full types in a little bit of the logging in
SendNonSendable where seeing the full type speeds up debugging in the context of
debugging user code and where in truth, I will never ever want to suppress types
and will not use this to create test cases. It is just for quick scanning.

gottesmm commented:
@swift-ci test

eeckstein commented on include/swift/SIL/SILPrintContext.h:
```diff
@@ -29,6 +29,27 @@ class SILBasicBlock;
+    PrintFullConvention = 0x8,
+
+    /// Print out all types that normally would be surpressed.
+    PrintAllTypes = 0x16,
```
`0x10`?

gottesmm replied:
You are right. Thanks! I was thinking 16 without hex.

eeckstein reviewed (approved):
The flag value seems to be wrong.
Otherwise lgtm

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
[sil] Apply SILPrintDebugInfo in SILPrinter instead of in SILPrinterContext.

SILPrinterContext is a currency type that is passed around and should be
transparent and not silently rely on a global variable. Now we instead allow
for the SILPrinter to override the behavior of the specified SILPrinterContext
using the global flag (matching the way needPrintTypeFor is implemented).

gottesmm committed:
[sil] Rename needPrintTypeFor -> shouldPrintType.

This makes it more naturally to read.

gottesmm commented:
@swift-ci smoke test
